### PR TITLE
Add all states

### PR DESCRIPTION
**Pull request by lcc**
Add all states that compose the knot-state-machine.

Depends on #21  .

### CODE REVIEW

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/base_classes/State.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package com.cesar.knot_sdk.knot_state_machine.states.base_classes
+
+abstract class State {
+
+
```
Comment (alla3) [inline knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/base_classes/State.kt]: Remove a blank line.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/base_classes/State.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    enum class ReadyStates{
+        REGISTER,
+        AUTHENTICATE
+    }
```
Comment (alla3) [inline knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/base_classes/State.kt]: The class state should not have this information.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/KNoTStateMachine.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    fun enterState() {
+        state.enter()
+    }
+
+    fun getUUUID() = sharedPref.getString(
+        UUID_PREF,
+        PREF_MISSING
+    )
+
+
```
Comment (alla3) [inline knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/KNoTStateMachine.kt:60]: Remove a blank line.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Unregister.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+class Unregister : State() {
+
+    override fun enter() {
+       TODO("Clear credentials")
+    }
+
+    override fun getNextState(event : KNoTEvent) = when(event) {
+        is NotReady        -> this
+        is Ready           -> this
```
Comment (alla3) [inline knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Unregister.kt:16]: Should the Unregister State act as a final state? IMO It should have a transition for the Disconnect State.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Disconnected.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        is RegNotOk        -> this
+        is AuthNotOk       -> this
+        is AuthOk          -> this
+        is SchemaOk        -> this
+        is SchemaNotOk     -> this
+        is UnregisterEvent -> this
+    }
+
+}
+
```
Comment (alla3) [inline knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Disconnected.kt]: Remove this blank line.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Registering.kt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        sharedPref.edit().putString(UUID_PREF, res.token).apply()
+        return TODO("Return schema state once it's been created")
+    }
+
+    override fun getNextState(event : KNoTEvent) = when(event) {
+        is NotReady        -> Disconnected()
+        is Ready           -> this
+        is Timeout         -> this
+        is RegOK           -> consumeMessage(event.res)
+        is RegNotOk        -> this
```
Comment (alla3) [inline knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Registering.kt:38]: IMO it should go to Error State if the event is RegNotOk.

**Review by alla3 — CHANGES_REQUESTED** [review]

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Unregister.kt (hunk 4 above)
Comment (lcc) [reply knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Unregister.kt:16]: That could be a future improvement.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Registering.kt (hunk 6 above)
Comment (lcc) [reply knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Registering.kt:38]: I agree, but I think this should be further discussed, for now, I am going to make this change.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Online.kt [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+        try {
+            knotMessager.publishData(updateData)
+        } catch (e : IOException) {
+            error = NETWORK_ERROR_MESSAGE
+        }
+    }
+
+
+
```
Comment (alla3) [inline knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Online.kt]: Remove two blank lines.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Registering.kt (hunk 6 above)
Comment (alla3) [reply knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Registering.kt:38]: Will you change it?

**Review by alla3 — CHANGES_REQUESTED** [review]

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Registering.kt (hunk 6 above)
Comment (alla3) [reply knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Registering.kt:38]: If you change it, make sure to update it on #23.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Online.kt (hunk 7 above)
Comment (lcc) [reply knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Online.kt]: Done.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Registering.kt (hunk 6 above)
Comment (lcc) [reply knot-thing/src/main/java/com/cesar/knot_sdk/knot_state_machine/states/Registering.kt:38]: For now, we only now that the register fails when the credentials are invalid (i.e.: someone owns the id ).Give this, I opted for generating a new ID and sending it again. We should probably look into other scenarios where register fails though.

**Review by alla3 — APPROVED** [review]

**Review by alla3 — CHANGES_REQUESTED** [review]
Solve the conflicting files.

**Review by alla3 — APPROVED** [review]

**Pull request merged**